src/sentry/api/endpoints/user_notification_fine_tuning.py

```diff
@@ -0,0 +1,199 @@
+            user,
+            request.user,
+            notifications,
+            notification_type=notification_type
+        )
+        return Response(serialized)
+
+    def put(self, request, user, notification_type):
+        if notification_type not in KEY_MAP:
+            return Response(status=status.HTTP_404_NOT_FOUND)
+
+        key = KEY_MAP[notification_type]
+        filter_args = {
+            'user': user,
+            'key': key['key'],
+        }
+
+        if notification_type == 'reports':
+            (user_option, created) = UserOption.objects.get_or_create(**filter_args)
+
+            value = user_option.value or []
+
+            # set of org ids that user is a member of
+            org_ids = self.get_org_ids(user)
+            for org_id, enabled in request.DATA.items():
+                org_id = int(org_id)
+                # We want "0" to be falsey
+                enabled = int(enabled)
+
+                # make sure user is in org
+                if org_id not in org_ids:
+                    return Response(status=status.HTTP_403_FORBIDDEN)
+
+                if enabled:
+                    value.remove(org_id)
+                else:
+                    value.insert(0, org_id)
+
+            user_option.update(value=value)
+            return Response(status=status.HTTP_204_NO_CONTENT)
+
+        if notification_type in ['alerts', 'workflow', 'email']:
+            update_key = 'project'
+            parent_ids = set(self.get_project_ids(user))
+        else:
+            update_key = 'organization'
+            parent_ids = set(self.get_org_ids(user))
+
+        ids_to_update = set([int(i) for i in request.DATA.keys()])
+
+        # make sure that the ids we are going to update are a subset of the user's
+        # list of orgs or projects
+        if not ids_to_update.issubset(parent_ids):
+            return Response(status=status.HTTP_403_FORBIDDEN)
+
+        if notification_type == 'email':
+            # make sure target emails exist and are verified
+            emails_to_check = set(request.DATA.values())
+            emails = UserEmail.objects.filter(
+                user=user,
+                email__in=emails_to_check,
+                is_verified=True
+            )
+
+            # Is there a better way to check this?
+            if len(emails) != len(emails_to_check):
+                return Response(status=status.HTTP_400_BAD_REQUEST)
+
+        with transaction.atomic():
+            for id in request.DATA:
+                val = request.DATA[id]
+                int_val = int(val) if notification_type != 'email' else None
+
+                filter_args['%s_id' % update_key] = id
+
+                # 'email' doesn't have a default to delete, and it's a string
+                # -1 is a magic value to use "default" value, so just delete option
+                if int_val == -1:
+                    try:
+                        UserOption.objects.get(**filter_args).delete()
```

Comment:
nit, i *think* it's more efficient to do `UserOption.objects.filter(**filter_args).delete()` since i think the above executes two queries (one to retrieve the row, another to delete) where as this one results in a single delete query. Then you also don't have to handle the `DoesNotExist` error

Reply:
It is more efficient, unless something depends on the model signals, because they don't get dispatched when doing a `filter(..).delete()`.